src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/3.diff.sql

```diff
@@ -0,0 +1,132 @@
+    IsPartiallySupported bit NOT NULL
+)
+
+CREATE UNIQUE CLUSTERED INDEX IXC_SearchParamRegistry
+ON dbo.SearchParamRegistry
+(
+    Uri
+)
+
+GO
+
+/*************************************************************
+    Stored procedures for the search parameter registry
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     Gets all the search parameters and their statuses.
+--
+-- DESCRIPTION
+--     Retrieves and returns the contents of the search parameter registry.
+--
+-- RETURN VALUE
+--     The search parameters and their statuses.
+--
+CREATE PROCEDURE dbo.GetSearchParamStatuses
+AS
+    SET NOCOUNT ON
+    
+    SELECT * FROM dbo.SearchParamRegistry
+GO
+
+--
+-- STORED PROCEDURE
+--     Given a table of search parameters, upserts the registry.
+--
+-- DESCRIPTION
+--     If a parameter with a matching URI already exists in the registry, it is updated.
+--     If not, a new entry is created.
+--
+-- PARAMETERS
+--     @searchParamStatuses
+--         * The updated or new search parameter statuses
+--
+CREATE PROCEDURE dbo.UpsertSearchParamStatus
+    @searchParamStatuses dbo.SearchParamRegistryTableType_1 READONLY
+AS
+    SET NOCOUNT ON
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    DECLARE @lastUpdated datetimeoffset(7) = SYSDATETIMEOFFSET()
+
+    -- Acquire and hold an exclusive table lock for the entire transaction to prevent parameters from being added or modified during upsertion.
+    UPDATE dbo.SearchParamRegistry
+    WITH (TABLOCKX)
+    SET Status = sps.Status, LastUpdated = @lastUpdated, IsPartiallySupported = sps.IsPartiallySupported
+    FROM dbo.SearchParamRegistry INNER JOIN @searchParamStatuses as sps
+    ON dbo.SearchParamRegistry.Uri = sps.Uri
+
+    INSERT INTO dbo.SearchParamRegistry
+        (Uri, Status, LastUpdated, IsPartiallySupported)
+    SELECT sps.Uri, sps.Status, @lastUpdated, sps.IsPartiallySupported
+    FROM @searchParamStatuses AS sps
+    WHERE sps.Uri NOT IN
+        (SELECT Uri FROM dbo.SearchParamRegistry) 
+
+    COMMIT TRANSACTION
+GO
+
+--
+-- STORED PROCEDURE
+--     Counts the number of search parameters.
+--
+-- DESCRIPTION
+--     Retrieves and returns the number of rows in the search parameter registry.
+--
+-- RETURN VALUE
+--     The number of search parameters in the registry.
+--
+CREATE PROCEDURE dbo.GetSearchParamRegistryCount
+AS
+    SET NOCOUNT ON
+    
+    SELECT COUNT(*) FROM dbo.SearchParamRegistry
+GO
+
+--
+-- STORED PROCEDURE
+--     Inserts a search parameter and its status into the search parameter registry.
+--
+-- DESCRIPTION
+--     Adds a row to the search parameter registry.
+--
+-- PARAMETERS
+--     @searchParamStatuses
+--         * The updated search parameter statuses
+--
+CREATE PROCEDURE dbo.InsertIntoSearchParamRegistry
+    @searchParamStatuses dbo.SearchParamRegistryTableType_1 READONLY
+AS
+    SET NOCOUNT ON
+
+    SET XACT_ABORT ON
+    BEGIN TRANSACTION
```

Comment:
Do we need transaction here since there is single insert operation?

Reply:
Good question. It depends on if we are using `lastUpdated` to manage concurrency on initialization. For now, I am wrapping `GetSearchParamRegistryCount` and `InsertIntoSearchParamRegistry` in a SQL transaction on the C# side, so I can remove it. Thanks for pointing it out 👍

Reply:
Fixed in c352d06.